### PR TITLE
Suport msgraph sdk in ansible

### PR DESCRIPTION
**Pull request by Fred-sun**
##### SUMMARY

Suport msgraph sdk in ansible.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

azure_rm_msuser.py
azure_rm_msuser_info.py
azure_rm_msgroup.py
azure_rm_msgroup_info.py
azure_rm_msserviceprincipal.py
azure_rm_msserviceprincipal_info.py
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/module_utils/azure_rm_common.py [hunk 1]
```diff
@@ -1451,6 +1452,12 @@ def datafactory_client(self):
     def datafactory_model(self):
         return DataFactoryModel
 
+    def get_msgraph_client(self):
+        credential = self.azure_auth.azure_credential_track2
+        client = GraphClient(credential=credential)
```
Comment (d2a-pnagel) [inline plugins/module_utils/azure_rm_common.py]: these 2 lines could be combined (`return GraphClient...`)

File: plugins/modules/azure_rm_msgroup.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                flag = False
+                for item in response:
+                    if item['displayName'] == self.display_name:
+                        flag = True
+                        response = item
+                        break
+                if not flag:
+                    response = None
+        except Exception as e:
+            self.log("There is not ad groups {0}".format(str(e)))
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msgroup.py]: There is no group / There are no groups

File: plugins/modules/azure_rm_msgroup.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+            else:
+                changed = True
+                response = self.delete_resource(response['id'])
+        else:
+            if self.state == 'present':
+                changed = True
+                response = self.create_resource(self.body)
+            else:
+                changed = False
+                self.log("The ad groups is not exist")
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msgroup.py]: The ad groups do not exist

File: plugins/modules/azure_rm_msgroup_info.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        description:
+            - The name of an attribute that you want to match to attribute_value.
+            - If attribute_name is not a collection type it will return groups where attribute_name is equal to attribute_value.
+            - If attribute_name is a collection type it will return groups where attribute_value is in attribute_name.
+        type: str
+    attribute_value:
+        description:
+            - The value to match attribute_name to.
+            - If attribute_name is not a collection type it will return groups where attribute_name is equal to attribute_value.
+            - If attribute_name is a collection type it will groups users where attribute_value is in attribute_name.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msgroup_info.py]: groups users?

File: plugins/modules/azure_rm_msgroup_info.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        type: str
+    all:
+        description:
+            - If True, will return all groups in tenant.
+            - If False will return no users.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msgroup_info.py]: no users -> no groups?

File: plugins/modules/azure_rm_msgroup_info.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    sample: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+displayName:
+    description:
+        - The display name of the group.
+    returned: always
+    type: str
+    sample: GroupName
+securityEnabled:
+    description:
+        - Whether the group is security-enable.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msgroup_info.py]: enabled

File: plugins/modules/azure_rm_msgroup_info.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+mail_enabled:
+    description:
+        - Whether the group is mail-enabled. Must be false.
+        - This is because only pure security groups can be created using the Graph API.
+    returned: always
+    type: bool
+    sample: False
+description:
+    description:
+        - Describe of the ad group.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msgroup_info.py]: description

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+    - Manage Azure Active Directory service principal.
+
+options:
+    app_id:
+        description:
+            - The application ID.
+        type: str
+    object_id:
+        description:
+            - It's service principal's object ID.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: It's -> The

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+        type: str
+    name:
+        description:
+            - The display name of the service principal.
+        type: str
+    web:
+        description:
+            - Redirects the Web URI,
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: "Web URI do redirect to." ? When does it redirect?

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+        type: dict
+        suboptions:
+            redirectUris:
+                description:
+                    - The authentication response to this URI.
+                type: list
+                elements: str
+    spa:
+        description:
+            - Redirects the single page application(SPA) URI,
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: "URI of the Single page application (SPA) to redirect to." ? When does it redirect?

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+        type: dict
+        suboptions:
+            redirectUris:
+                description:
+                    - The authentication response to this URI.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: This does not make sense to me.  `redirectUris` = a response?

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+            - The display name of the service principal.
+        type: str
+    web:
+        description:
+            - Redirects the Web URI,
+        type: dict
+        suboptions:
+            redirectUris:
+                description:
+                    - The authentication response to this URI.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: This does not make sense to me.  `redirectUris` = a response? Do you mean "The URI to redirect to, after authentication."?

File: tests/integration/targets/azure_rm_msserviceprincipal/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+- name: Create microsoft graph service principal
+  azure_rm_msserviceprincipal:
+    name: "{{ rpfx }}"
+    sign_in_audience: AzureADandPersonalMicrosoftAccount
+    public_client:
+      redirectUris:
+        - https://localhost
+    spa:
+      redirectUris:
+        - https://localhost11
```
Comment (d2a-pnagel) [inline tests/integration/targets/azure_rm_msserviceprincipal/tasks/main.yml]: 11, not 1?

File: plugins/modules/azure_rm_msuser_info.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+                for item in response:
+                    if item[self.attribute_name] == self.attribute_value:
+                        ms_user.append(item)
+            elif self.all:
+                ms_user = client.get('/users/').json()['value']
+
+            self.results['ms_users'] = [self.to_dict(user) for user in ms_user]
+
+        except Exception as e:
+            self.fail("failed to get Microsoft Grpah user info {0}".format(str(e)))
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msuser_info.py]: Graph

File: plugins/modules/azure_rm_msuser_info.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    - name: Using object_id
+      azure.azcollection.azure_rm_msaduser_info:
+        object_id: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+
+    - name: Using attribute mail - not a collection
+      azure.azcollection.azure_rm_msuser_info:
+        attribute_name: mail
+        attribute_value: foo@contoso.com
+
+    - name: Get all user in same tenant
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msuser_info.py]: Users

File: plugins/modules/azure_rm_msuser_info.py [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+
+short_description: Get Azure Microsoft Graph user info
+
+description:
+    - Get Microsoft Graph user info.
+
+options:
+    object_id:
+        description:
+            - The object id for the user.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msuser_info.py:23]: "object id" / "object ID" (inconsistent casing)

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+                response = self.create_resource(self.body)
+            else:
+                changed = True
+                response = self.delete_resource(response['id'])
+        else:
+            if self.state == 'present':
+                response = self.create_resource(self.body)
+                changed = True
+            else:
+                self.log("The Service principal is not exist")
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: does not exist

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+                for item in response:
+                    if item['appId'] == self.app_id:
+                        flag = True
+                        response = item
+                        break
+                if not flag:
+                    response = None
+
+        except Exception as e:
+            self.log("There is not service principal {0}".format(str(e)))
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: not -> no

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+        type: dict
+        suboptions:
+            redirectUris:
+                description:
+                    - The authentication response to this URI.
+                type: list
+                elements: str
+    public_client:
+        description:
+            - Redirects the public client/native URI,
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: same remark as above

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+        type: dict
+        suboptions:
+            redirectUris:
+                description:
+                    - The authentication response to this URI.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: same remark as above

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+    sample: []
+object_id:
+    description:
+        - Object ID of the associated service principal.
+    returned: always
+    type: str
+    sample: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+spa:
+    description:
+        - Redirects the single page application(SPA) URI,
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: same remark as above

File: plugins/modules/azure_rm_msserviceprincipal.py [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+    returned: always
+    type: dict
+    sample: {'redirectUris':['https://spa.com']}
+web:
+    description:
+        - Redirects the Web URI,
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal.py]: same remark as above

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+    - Get Azure Active Directory service principal info.
+
+options:
+    app_id:
+        description:
+            - The application ID.
+        type: str
+    object_id:
+        description:
+            - It's service principal's object ID.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: It's -> The

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+        type: str
+    all:
+        description:
+            - If True, will return all applicatin in tenant.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: applications

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+extends_documentation_fragment:
+    - azure.azcollection.azure
+
+author:
+    - xuzhang3 (@xuzhang3)
+    - Fred-sun (@Fred-sun)
+'''
+
+EXAMPLES = '''
+  - name: get MS sp info by object_id
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: sp -> service provider

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+    azure_rm_msserviceprincipal_info:
+      object_id: "{{ object_id }}"
+
+  - name: get all MS sp info
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: sp -> service provider

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+    azure_rm_msserviceprincipal_info:
+        all: True
+
+  - name: get MS sp info by App ID
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: sp -> service provider

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+    sample: sp
+app_id:
+    description:
+        - The application ID.
+    returned: always
+    type: str
+    sample: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+app_roles:
+    description:
+        -  The collection of application roles that an application may declare.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: extra space

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+        - These roles can be assigned to users, groups or service principals.
+    type: list
+    returned: always
+    sample: []
+object_id:
+    description:
+        - It's service principal's object ID.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: It's -> The

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+    returned: always
+    type: str
+    sample: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+spa:
+    description:
+        - Redirects the single page application(SPA) URI,
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: same remark as earlier on spa/description

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+    returned: always
+    type: dict
+    sample: {'redirectUris':['https://spa.com']}
+web:
+    description:
+        - Redirects the Web URI,
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: same remark as earlier on web/description

File: plugins/modules/azure_rm_msserviceprincipal_info.py [hunk 32]
```diff
@@ -0,0 +1,181 @@
...
+    returned: always
+    type: dict
+    sample: {'redirectUris':['https://web.com']}
+public_client:
+    description:
+        - Redirects the public client/native URI.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msserviceprincipal_info.py]: same remark as earlier

File: plugins/modules/azure_rm_msuser.py [hunk 33]
```diff
@@ -0,0 +1,283 @@
...
+    sample: 824736848qq.onmicrosoft.com
+surname:
+    description:
+        - The user's surname (family name or last name).
+    type: str
+    returned: always
+    sample: '张'
+givenName:
+    description:
+        - The given name for the user.
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msuser.py]: of the user

File: plugins/modules/azure_rm_msuser.py [hunk 34]
```diff
@@ -0,0 +1,283 @@
...
+                response = self.delete_resource(response['id'])
+                changed = True
+        else:
+            if self.state == 'present':
+                response = self.create_resource(self.body)
+                changed = True
+            else:
+                changed = False
+                response = None
+                self.log("The ad user account not exist")
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msuser.py]: does not exist

**Review by d2a-pnagel — CHANGES_REQUESTED** [review]
Sorry, I only have remarks on the descriptions / comments. Hope it helps.

**Review by d2a-pnagel — CHANGES_REQUESTED** [review]
Sorry, I only have remarks on the descriptions / comments. Hope it helps.

File: tests/integration/targets/azure_rm_msserviceprincipal/tasks/main.yml (hunk 13 above)
Comment (Fred-sun) [reply tests/integration/targets/azure_rm_msserviceprincipal/tasks/main.yml]: just a case, ignore this!

**Comment by Fred-sun** [comment]
> Sorry, I only have remarks on the descriptions / comments. Hope it helps.

Update! Thanks for your review！

File: plugins/modules/azure_rm_msgroup.py [hunk 35]
```diff
@@ -0,0 +1,266 @@
...
+                flag = False
+                for item in response:
+                    if item['displayName'] == self.display_name:
+                        flag = True
+                        response = item
+                        break
+                if not flag:
+                    response = None
+        except Exception as e:
+            self.log("There is not no group {0}".format(str(e)))
```
Comment (d2a-pnagel) [inline plugins/modules/azure_rm_msgroup.py]: "not no" -> "no"

**Review by d2a-pnagel — DISMISSED** [review]

**Comment by sboulkour** [comment]
Hello @Fred-sun !
Thx for your work on this, much appreciated :)

I have two question though :
- are resources renamed from ad* to ms* to mark the switch from AD Graph to MS Graph ?
- and in the listed resources I see some missing that would be of use for m personnal use case (azure_rm_adapplication), is it possible to add it to your current PR ?

**Comment by Fred-sun** [comment]
@sboulkour Yes, Graph RBAC will be migrated to MS Graph and we are adding azure_rm_adapplication and azure_rm_adpass modules! Thanks!

**Comment by kmj251** [comment]
Will this PR be merged, and an updated version of the collection released before Azure AD Graph is turned off on June 30, 2023?

**Comment by Fred-sun** [comment]
Currently, msgraph-core cannot be directly used for aad management, and we are analyzing other ways to solve it. Shut it down for now.

**Comment by brootware** [comment]
Hello @Fred-sun, came across this thread and found out that this was not merged for Azure b2b collaboration. We ran into a use case to automate this part. May I know if there has been any development/update on this?

**Comment by Fred-sun** [comment]
@sboulkour Has finished, Please reference https://github.com/ansible-collections/azure/pull/1325

**Pull request closed without merging**